File: test/test_isa.cpp
#include <iostream>
#include <cassert>
#include "hardware/isa.hpp"

int main() {
    Instruction orig;
    orig.opcode    = OP_MAT_LOAD;
    orig.sram_bank = 1;
    orig.dram_addr = 0x1234;
    orig.sram_addr = 0x5A;

    uint32_t packed = orig.encode();
    std::cout << "Packed 32-bit Hex: 0x" << std::hex << packed << std::endl;

    Instruction unpacked = Instruction::decode(packed);

    assert(unpacked.opcode == orig.opcode);
    assert(unpacked.sram_bank == orig.sram_bank);
    assert(unpacked.dram_addr == orig.dram_addr);
    assert(unpacked.sram_addr == orig.sram_addr);

    std::cout << "SUCCESS: All ISA bits encoded and decoded perfectly!" << std::endl;
    return 0;
}

File: test/test_isa.py
# test/test_isa.py
from compiler.isa import Instruction, OP_MAT_LOAD

def test_isa_encoding():
    # Same values we tested in C++: OP_MAT_LOAD, bank=1, dram=0x1234, sram=0x5A
    inst = Instruction(OP_MAT_LOAD, 1, 0x1234, 0x5A)
    packed = inst.encode()
    print(f"Python Packed Hex: {hex(packed)}")
    
    # Assert exact match with C++ test result (0x312345A)
    assert packed == 0x312345A, f"Expected 0x312345A, got {hex(packed)}"

    # Test decode
    unpacked = Instruction.decode(packed)
    assert unpacked.opcode == OP_MAT_LOAD
    assert unpacked.sram_bank == 1
    assert unpacked.dram_addr == 0x1234
    assert unpacked.sram_addr == 0x5A
    print("SUCCESS: Python ISA Encoder & Decoder verified against C++!")

if __name__ == "__main__":
    test_isa_encoding()